src/include/duckdb/common/arrow/appender/varchar_data.hpp

```diff
@@ -112,4 +113,65 @@ struct ArrowVarcharData {
 	}
 };
 
+struct ArrowVarcharToStringViewData {
+	static void Initialize(ArrowAppendData &result, const LogicalType &type, idx_t capacity) {
+		result.main_buffer.reserve((capacity + 1) * sizeof(arrow_string_view_t));
+		result.aux_buffer.reserve(capacity);
+	}
+
+	static void Append(ArrowAppendData &append_data, Vector &input, idx_t from, idx_t to, idx_t input_size) {
+		idx_t size = to - from;
+		UnifiedVectorFormat format;
+		input.ToUnifiedFormat(input_size, format);
+
+		// resize the validity mask and set up the validity buffer for iteration
+		ResizeValidity(append_data.validity, append_data.row_count + size);
+		auto validity_data = (uint8_t *)append_data.validity.data();
+
+		// resize the offset buffer - the offset buffer holds the offsets into the child array
+		auto data = UnifiedVectorFormat::GetData<string_t>(format);
+		idx_t last_offset = 0;
+		for (idx_t i = from; i < to; i++) {
+			auto result_idx = append_data.row_count + i - from;
+			auto arrow_data = append_data.main_buffer.GetData<arrow_string_view_t>();
+			auto source_idx = format.sel->get_index(i);
+			if (!format.validity.RowIsValid(source_idx)) {
+				// Null value
+				uint8_t current_bit;
+				idx_t current_byte;
+				GetBitPosition(result_idx, current_byte, current_bit);
+				SetNull(append_data, validity_data, current_byte, current_bit);
+				continue;
+			}
+			// These two are now the same buffer
+			idx_t string_length = ArrowVarcharConverter::GetLength(data[source_idx]);
+			auto string_data = data[source_idx].GetData();
+			if (string_length <= arrow_string_view_t::max_inlined_bytes) {
+				//	This string is inlined
+				//  | Bytes 0-3  | Bytes 4-15                            |
+				//  |------------|---------------------------------------|
+				//  | length     | data (padded with 0)                  |
+				arrow_data[result_idx] = arrow_string_view_t(string_length, string_data);
+			} else {
+				// This string is not inlined, we have to check a different buffer and offsets
+				//  | Bytes 0-3  | Bytes 4-7  | Bytes 8-11 | Bytes 12-15 |
+				//  |------------|------------|------------|-------------|
+				//  | length     | prefix     | buf. index | offset      |
+				arrow_data[result_idx] = arrow_string_view_t(string_length, string_data, 0, last_offset);
+				auto current_offset = last_offset + string_length;
+				append_data.aux_buffer.resize(current_offset);
+				memcpy(append_data.aux_buffer.data() + last_offset, string_data, string_length);
+				last_offset = current_offset;
+			}
+		}
+		append_data.row_count += size;
+	}
+
+	static void Finalize(ArrowAppendData &append_data, const LogicalType &type, ArrowArray *result) {
+		result->n_buffers = 3;
+		result->buffers[1] = append_data.main_buffer.data();
+		result->buffers[2] = append_data.aux_buffer.data();
```

Comment:
For binary or utf-8 view arrays, an extra buffer is appended which stores
the lengths of each variadic data buffer as `int64_t`. This buffer is
necessary since these buffer lengths are not trivially extractable from
other data in an array of binary or utf-8 view type.

```suggestion
		result->n_buffers = 4;
		result->buffers[1] = append_data.main_buffer.data();
		result->buffers[2] = append_data.aux_buffer.data();
```
https://github.com/apache/arrow/blob/d1fa0b1cc6321d2dc2ec16608faa8bfc8e2607be/docs/source/format/CDataInterface.rst?plain=1#L555-L561

Reply:
Ah, I think I missed that in the docs, so for each data buffer we also should append and populate an offset buffer?

Reply:
Not an extra buffer for _each_ data buffer, but just one extra buffer that contains the lengths (not offsets) of the data buffers.

Reply:
For a code example, see the `variadic_buffer_sizes_` in the Arrow C++ implementation (https://github.com/apache/arrow/blob/91bf1c9c170c1917ad47bb0dbb38aa5c9fbbbfb2/cpp/src/arrow/c/bridge.cc#L537 / https://github.com/apache/arrow/blob/91bf1c9c170c1917ad47bb0dbb38aa5c9fbbbfb2/cpp/src/arrow/c/bridge.cc#L595-L599)

Reply:
Aha, got it, thanks for the clarification!

Reply:
So to clarify, the last buffer of the string view array contains the lengths of the buffers ?

Reply:
Yes, indeed!
If any wording in https://arrow.apache.org/docs/dev/format/CDataInterface.html#binary-view-arrays can be changed to make this clearer, suggestions welcome (maybe making it more visual like your example might help)

Reply:
@jorisvandenbossche, I have a question related to data_buffers: Is it possible to create an Arrow table from Python using string_view objects that span multiple data buffers?

Reply:
Not directly manually with a small toy example (I notice that `pa.Array.from_buffers` doesn't yet work for string_view, because there is a hardcoded check for the number of buffers that need to be updated). 

But you can create such an array by creating it with more data (the underlying builder chunks the data buffers with a default blocksize of 32KB):

```
In [1]: arr = pa.array(["Imaverybigstringmuchbiggerthanfourbytes"], type=pa.string_view())

In [2]: arr.buffers()
Out[2]: 
[None,
 <pyarrow.Buffer address=0x7f19c900a040 size=16 is_cpu=True is_mutable=True>,
 <pyarrow.Buffer address=0x7f19c900a080 size=39 is_cpu=True is_mutable=True>]

In [3]: arr = pa.array(["Imaverybigstringmuchbiggerthanfourbytes"]*5000, type=pa.string_view())

In [4]: arr.buffers()
Out[4]: 
[None,
 <pyarrow.Buffer address=0x7f19c90ae3c0 size=80000 is_cpu=True is_mutable=True>,
 <pyarrow.Buffer address=0x7f19c90c8340 size=32760 is_cpu=True is_mutable=True>,
 <pyarrow.Buffer address=0x7f19c90d1280 size=32760 is_cpu=True is_mutable=True>,
 <pyarrow.Buffer address=0x7f19c90da140 size=32760 is_cpu=True is_mutable=True>,
 <pyarrow.Buffer address=0x7f19c90e33c0 size=32760 is_cpu=True is_mutable=True>,
 <pyarrow.Buffer address=0x7f19c90ec200 size=32760 is_cpu=True is_mutable=True>,
 <pyarrow.Buffer address=0x7f19c90f51c0 size=31200 is_cpu=True is_mutable=True>]
```

Alternatively (and this might be more robust in case pyarrow would change that default blocksize), you can also concatenate two small arrays, which will only physically concatenate the validity and views buffers, and just take the original data buffers:

```
In [7]: arr = pa.array(["Imaverybigstringmuchbiggerthanfourbytes"], type=pa.string_view())

In [9]: pa.concat_arrays([arr, arr]).buffers()
Out[9]: 
[None,
 <pyarrow.Buffer address=0x7f19c900a140 size=32 is_cpu=True is_mutable=True>,
 <pyarrow.Buffer address=0x7f19c900a100 size=39 is_cpu=True is_mutable=True>,
 <pyarrow.Buffer address=0x7f19c900a100 size=39 is_cpu=True is_mutable=True>]
```

Reply:
> Perhaps a reference to that section from here https://arrow.apache.org/docs/format/Columnar.html#variable-size-binary-view-layout would be good, that's where we went to look for the spec

Good idea! Doing an attempt to clarify this at https://github.com/apache/arrow/pull/40156